Add support for Directory.Packages.props file as entrypoint

TobiasLaving opened the pull request:
This PR will enable dependabot to use a Directory.Packages.props file as entry point, allowing the setup below:

Repo:
```
src/
├─ Backend/
│  ├─ Project1/
│  │  ├─ Project1.csproj
│  ├─ Project2/
│  │  ├─ Project2.csproj
│  ├─ Directory.Packages.props
├─ App/
├─ Solution.sln
```

dependabot.yml:
```

version: 2
updates:
  - package-ecosystem: "nuget"
    directory: "/src/Backend"
    schedule:
      interval: "daily"
    open-pull-requests-limit: 5
    target-branch: master
```


Note: I have never written ruby before and I'm a bit out of my depth
This is working locally towards our repo, but I'm guessing there is a bunch of things I am missing.
I'm happy to get any feedback

yeikel commented:
Please add tests for this change

TobiasLaving commented:
The issue is discussed in the issue: https://github.com/dependabot/dependabot-core/issues/5635

I will look into adding tests!
On another note, as seen in the PR two files have identical functions, is this intentional?
Should the function "project_files" be broken out?

yeikel commented:
> The issue is discussed in the issue: https://github.com/dependabot/dependabot-core/issues/5635
> 
> I will look into adding tests!
> On another note, as seen in the PR two files have identical functions, is this intentional?
> Should the function "project_files" be broken out?

Refactoring it sounds reasonable to me

TobiasLaving commented:
I now added some tests.

I used this repo to create the [fixtures](https://github.com/TobiasLaving/nuget-directory-packages-props-example)

The repo has drifted from what the PR originally described and looks more like this:
```
[tobias@tobias-expertbook nuget-directory-packages-props-example]$ tree -P "*.sln|*.csproj|Directory.Packages.props" --prune
.
├── App1
│   ├── App1.sln
│   └── src
│       └── Backend
│           ├── Project1
│           │   └── Project1.csproj
│           └── Project2
│               └── Project2.csproj
└── Directory.Packages.props
```

The intended dependabot.yml would look like this:
```
version: 2
updates:
  - package-ecosystem: "nuget"
    directory: "/"
    schedule:
      interval: "daily"
    open-pull-requests-limit: 5
    target-branch: master
```

The reason for the drift is entirely selfish, it is more like our private repository. Still, the repo in the PR should also work.

For testing I added 
1. Simple test for getting only directory.packages.props if it is found in the root
2. Test to make sure the new code does not override the old logic for sln and all files are still fetched.

I wanted to do some refactoring but quickly felt out of my depth so I fell back onto doing the least work possible.
Happy to hear any feedback, in particular it would be helpful if we can figure out more tests cases.

jeffwidman commented:
@TobiasLaving just to clarify, did you see 2️⃣ in my comment above? Can you add those two simple tests?

DkSkydancer commented:
@TobiasLaving What is the status on this?
@jeffwidman How long we should expect this PR to hold of for https://github.com/dependabot/dependabot-core/pull/6031

TobiasLaving commented:
@DkSkydancer This PR is ready in my opinion

DkSkydancer commented:
@TobiasLaving  @jeffwidman  ping. we are still waiting for this!

TobiasLaving commented:
Seems @jeffwidman is out (?)
Do you know the next steps @yeikel ?

Nishnha commented:
Hey folks, Dependabot maintainers here - we've had a bit of a deploy freeze and have fallen behind on reviews. Sorry about that!

I took a look at the changes and I agree that this PR is in a good state. We should be able to take a look at merging it some time this week. 

In the meantime, would you be able to bring it up-to-date with the base branch @TobiasLaving? The CI tests should be able to run now

DkSkydancer commented:
@TobiasLaving Can you the update as requested above pl?

TobiasLaving commented:
@DkSkydancer @Nishnha 
PR is now rebased and all checks has passed :+1:

DkSkydancer commented:
@Nishnha the PR should be ready now

DkSkydancer commented:
@Nishnha  @jeffwidman Can we get a maintainer review please?

DkSkydancer commented:
@TobiasLaving it looks like this branch is out of date again I'm afraid :(

Nishnha commented on nuget/lib/dependabot/nuget/file_parser.rb:
```diff
@@ -70,7 +70,7 @@ def project_file_parser
 
       def project_files
-        dependency_files.select { |df| df.name.match?(/\.[a-z]{2}proj$/) }
+        dependency_files.select { |df| df.name.match?(/\.[a-z]{2}proj$|[Dd]irectory.[Pp]ackages.props/) }
```
Could we split this regex into two different captures and try to match against both?

Something like 

```suggestion
        projfile = /\.[a-z]{2}proj$/
        packageprops = /[Dd]irectory.[Pp]ackages.props/
        
        dependency_files.select { |df| 
          df.name.match?(projfile) ||  
          df.name.match?(packageprops)
        }
```

and do the same in the file updater?

Nishnha commented on nuget/spec/dependabot/nuget/file_fetcher_spec.rb:
```diff
@@ -773,4 +828,37 @@
+        )
+    end
+
+    it "fetches the files the packages props file" do
```
typo
```suggestion
    it "fetches the packages props file" do
```

Nishnha commented on nuget/spec/fixtures/github/contents_dotnet_directory_packages_props.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "Directory.Packages.props",
```
For anyone else reviewing, this response is for the file https://github.com/TobiasLaving/nuget-directory-packages-props-example/blob/main/Directory.Packages.props

Nishnha commented:
Thank you for this contribution! I left a comment on the regex capture that I would like to see addressed before approving.

We can bring this PR up-to-date before we merge it, so it's fine that it has become out-of-date with the base branch.

@jeffwidman should we fork https://github.com/TobiasLaving/nuget-directory-packages-props-example into the dependabot org since it's referenced in the specs?

TobiasLaving commented:
Good points, I have made the changes locally sadly the dev container seems to be broken for nuget atm so I cannot run it locally: https://github.com/dependabot/dependabot-core/issues/7438

Will push once I can validate locally

TobiasLaving commented:
@Nishnha

Fixed your comments, thanks for finding and fixing the docker issue!

TobiasLaving commented:
@Nishnha There, the linter is now happy :+1: 

Struggeled a little bit with running the lint script in the dev container, sorry about the delay

TobiasLaving commented:
@Nishnha 
For my understanding, who does the actual merge?

Nishnha commented:
We have a deploy-then-merge process so it's not as straightforward as pressing the merge button. That said, this PR has been merged now 😄 

Thank you for the contribution!

jdwalker-emis commented:
@Nishnha hi is there please any way of tracking when this will go to full release across Github?
